swiftsimio/initial_conditions/generate_particles.py

```diff
@@ -1,5 +1,1047 @@
+        ``generate_IC_for_given_density()``
+
+
+        Parameters
+        -------------
+
+        boxsize: unyt_array, optional
+            The box size of the simulation.
+
+        periodic: bool, optional
+            whether the simulation box is periodic or not
+
+        nx: int, optional
+            how many particles along every dimension you want your simulation
+            to contain
+
+        ndim: int, optional
+            how many dimensions you want your simulation to have
+
+        unit_l:  None or unyt.unit_object.Unit or str, optional
+            unit length for the coordinates. If None, the same unit given for
+            the boxsize will be used.
+
+        unit_m: unyt.unit_object.Unit or str, optional
+            unit mass for the particle masses
+
+        kernel: str {'cubic spline',}, optional
+            which kernel to use
+
+        eta: float, optional
+            resolution eta, which defines the number of neighbours used
+            independently of dimensionality
+
+
+        Returns
+        ------------
+
+        ic_sim_params: dict
+            dict containing the parameters stored in a way such that ``generate_IC_for_given_density()``
+            understands them.
+
+
+        Notes
+        -----------
+        
+        + The returned dict is a required argument to call ``generate_IC_for_given_density()``
+        """
+
+        if not isinstance(boxsize, unyt.unyt_array):
+            raise TypeError("boxsize needs to be a unyt array.")
+
+        self.boxsize = boxsize
+        self.periodic = periodic
+        self.nx = nx
+        self.ndim = ndim
+        if unit_l is None:
+            self.unit_l = boxsize.units
+            self.boxsize_to_use = boxsize.value  # use only np.array
+        else:
+            self.unit_l = unit_l
+            self.boxsize_to_use = boxsize.to(unit_l).value
+        self.unit_m = unit_m
+        self.kernel = kernel
+        self.eta = eta
+        return
+
+    def copy(self):
+        r"""
+        Returns a copy of itself.
+        """
+        new = ic_sim_params(
+            boxsize=self.boxsize,
+            periodic=self.periodic,
+            nx=self.nx,
+            ndim=self.ndim,
+            unit_l=self.unit_l,
+            unit_m=self.unit_m,
+            kernel=self.kernel,
+            eta=self.eta,
+        )
+        return new
+
+
+class ic_run_params(object):
+    r"""
+    Objects containing run parameters for IC generation
+    """
+    iter_max: int = 2000
+    converge_thresh: float = 1e-4
+    tolerance_part: float = 1e-3
+    displ_thresh: float = 1e-3
+    delta_init: Union[float, None] = None
+    delta_reduct: float = 1.0
+    delta_min: float = 1e-6
+    redist_freq: int = 20
+    redist_frac: float = 0.01
+    redist_reduct: float = 1.0
+    no_redistribution_after: int = 200
+    dump_basename: str = "IC-generation-iteration-"
+    dumpfreq: int = 50
+
+    def __init__(
```

Comment:
same comments as above for your `__init__`.

Reply:
'tis done.